horovod/common/operations.cc

```diff
@@ -75,6 +108,63 @@ namespace horovod {
 namespace common {
 
 namespace {
+    
+#if HAVE_MLSL
+#define COMMUNICATOR_TABLE_MAX 1000
+#define NULL_VAL nullptr
+
+#define ERROR 0
+#define INFO  1
+#define DEBUG 2
+#define TRACE 3
+
+#define GET_TID() syscall(SYS_gettid)
+
+#define MLSL_LOG(log_lvl, fmt, ...)					\
+  do {									\
+    if (log_lvl <= mlsl_log_lvl)					\
+    {									\
+       char time_buf[20]; /*2016:07:21 14:47:39*/			\
+       GetTime(time_buf, 20);						\
+       switch (log_lvl) {						\
+       case ERROR:							\
+         printf("%s: ERROR: (%ld): %s:%u " fmt "\n", time_buf, GET_TID(),\
+            __FUNCTION__, __LINE__, ##__VA_ARGS__);			\
+         break;							\
+       case INFO:							\
+         printf("(%ld): %s:%u " fmt "\n", GET_TID(),			\
+            __FUNCTION__, __LINE__, ##__VA_ARGS__);			\
+         break;							\
+       case DEBUG:							\
+       case TRACE:							\
+         printf("%s: (%ld): %s:%u " fmt "\n", time_buf, GET_TID(),	\
+            __FUNCTION__, __LINE__, ##__VA_ARGS__);			\
+       default:							\
+         assert(0);							\
+       }								\
+       fflush(stdout);							\
+    }									\
+  } while (0)
+
+void GetTime(char* buf, size_t bufSize)
+{
+   time_t timer;
+   struct tm* timeInfo = 0;
+   time(&timer);
+   timeInfo = localtime(&timer);
+   assert(timeInfo);
+   strftime(buf, bufSize, "%Y:%m:%d %H:%M:%S", timeInfo);
+}
+
+enum HvdRequestType
```

Comment:
We already have request type in `MPIRequest::RequestType`, can we use that?

Reply:
@alsrgv Yes, we can. Done